Moving hooks to the packages they support

bjones1 opened the pull request:
This is an implementation of the ideas discussed in #4232. This is somewhat a work in progress, in the sense that some of the code probably belongs in another repository. The repository can serve as an example for package authors on how to include PyInstaller tests for their package.

I didn't bother supporting Python 2.7, since it's going away soon and I'm assuming PyInstaller 4.0 won't support it. The Appveyor tests "fail" (only on 2.7); Travis doesn't, since I eliminated 2.7 tests from that test suite.

@htgoebel, I'd like your feedback before proceeding. Thanks for your ideas along the way -- I think this could be a big step forward for PyInstaller. If we can get package authors to take on most of the responsibility for maintaining hooks, I think everyone will benefit.

Closes #4232

htgoebel commented on PyInstaller/utils/run_tests.py:
```diff
@@ -0,0 +1,36 @@
+
+# Run pytest.
+def run_pytest(*args, **kwargs):
+    pytest.main([*paths_to_test(), *args], **kwargs)
```
Assume I'm developing a hook for `scipy`, which requires `numpy` (not sure it it does, just let's assume it). Then this will find all `scipy` and `numpy` tests. But I only want to run the these in `scipy/tests/hooks/part3/`. How can I do this?

(This will always run *all* tests in all paths. User are not able to restrict this to a given directory any more.)

htgoebel replied:
```suggestion
    return pytest.main([*paths_to_test(), *args], **kwargs)
```

bjones1 replied:
Users could pass additional parameters of `["-k", "<specific tests to run>"]`. See https://docs.pytest.org/en/latest/usage.html#calling-pytest-from-python-code.

bjones1 replied:
I think I understand now -- I added a way to select only tests from a given module/submodule.

htgoebel commented on setup.py:
```diff
@@ -180,7 +180,7 @@ def run(self):
         # This file is necessary for rthooks (runtime hooks).
         'PyInstaller.loader': ['rthooks.dat'],
-        },
+    },
```
No white-space change, please.

bjones1 replied:
It looks like you fixed this during the rebase.

htgoebel commented on setup.py:
```diff
@@ -192,5 +192,27 @@ def run(self):
+            # strings, each of which is an absolute path to a directory tree to
+            # pass to pytest for test discovery or an absolute path to a Python
+            # source file to pass to pytest for test discovery.
+            'test = PyInstaller.utils.run_tests:PyInstaller_tests'
```
The comment could go into the manual. Thus here we only need something like "Make PyInstaller's hooks tests available, see ... manual".

htgoebel replied:
Should this be `test` or `tests` (plural) (I'm in favor of plural)

bjones1 replied:
I agree on `tests`. Updated. I'll probably use CodeChat to create a code-readable manual from the hooks repo you created.

htgoebel commented on setup.py:
```diff
@@ -192,5 +192,27 @@ def run(self):
+
+    # Requirements for running tests using the PyInstaller pytest framework.
+    extras_require={
+        'test': ['pytest >= 2.7.3'],
```
This should be reflected in the respective requirements file, too.

bjones1 replied:
I agree -- I actually coped it from `tests\requirements-tools.txt`.

htgoebel commented:
@bjones: Please find
- an updated branch, rebased on current develop-branch and Python 2 support removed at https://github.com/htgoebel/pyinstaller/tree/4232a
- an example package using this mechanism (and the aforementioned branch) at <https://github.com/pyinstaller/hooksample>

bjones1 commented:
@htgoebel, thanks! I'll work on this when I can...my workload is high, currently.

bjones1 commented on PyInstaller/depend/analysis.py:
```diff
@@ -133,6 +133,68 @@ def _reset(self, user_hook_dirs):
+                continue
+            except Exception as e:
+                logger.error('Unable to read run-time hooks from %s: %s' %
+                             (uhd_path, str(e)))
```
Thanks, fixed.

bjones1 commented on PyInstaller/depend/analysis.py:
```diff
@@ -133,6 +133,68 @@ def _reset(self, user_hook_dirs):
+            # The expected data structure for a run-time hook file is::
+            #
+            #   {
+            #       str module_name1: [
```
OK, updated.

bjones1 commented on PyInstaller/depend/analysis.py:
```diff
@@ -133,6 +133,68 @@ def _reset(self, user_hook_dirs):
+                        (uhd_path, module_name, python_file_name)
+                    )
+                    # Merge it.
+                    self._available_rthooks[module_name].append(abs_path)
```
Done.

bjones1 commented on PyInstaller/depend/analysis.py:
```diff
@@ -133,6 +133,68 @@ def _reset(self, user_hook_dirs):
+        # Search for run-time hooks in all hook directories.
+        self._available_rthooks = defaultdict(list)
+        for uhd in self._user_hook_dirs:
+            uhd_path = os.path.join(uhd, 'rthooks.dat')
```
Why? I don't see that this would affect the operation of the code -- whether `uhd` is relative or absolute, `open` will still look for it. Am I missing something?

bjones1 replied:
On second thought, I added it.

bjones1 commented on PyInstaller/depend/analysis.py:
```diff
@@ -133,6 +133,68 @@ def _reset(self, user_hook_dirs):
+            for module_name, python_file_name_list in available_rthooks.items():
+                # Ensure the key is a string.
+                assert isinstance(module_name, compat.text_type), (
+                    '%s must be a dict of keys; the key "%s" '
```
Thanks -- nice catch on the `%r`.

bjones1 commented on PyInstaller/depend/analysis.py:
```diff
@@ -133,6 +133,68 @@ def _reset(self, user_hook_dirs):
+                )
+                # Ensure the value is a list.
+                assert isinstance(python_file_name_list, list), (
+                    'The value of %s key "%s" must be a list.' % (uhd_path, module_name)
```
Good idea, but I'll postpone that. For now, I've decided to use `%r` for file names (which may have spaces/unusual characters) and `%s` for module names (which cannot).

bjones1 commented on PyInstaller/depend/analysis.py:
```diff
@@ -133,6 +133,68 @@ def _reset(self, user_hook_dirs):
+                for python_file_name in python_file_name_list:
+                    # Ensure each item in the list is a string.
+                    assert isinstance(python_file_name, compat.text_type), (
+                        '%s key "%s", item "%s" must be a string.' %
```
Done.

bjones1 commented on PyInstaller/utils/run_tests.py:
```diff
@@ -0,0 +1,36 @@
+# for distributing bootloader.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+#-----------------------------------------------------------------------------
```
Updated.

bjones1 commented on PyInstaller/utils/run_tests.py:
```diff
@@ -0,0 +1,36 @@
+    return [os.path.abspath(os.path.join(os.path.dirname(__file__), '../../tests'))]
+
+
+# Return a list of all paths to PyInstller-testable packages.
```
IMHO, docstrings are poor style -- documentation about a function should go before the function's name, not after it.

bjones1 commented on PyInstaller/utils/run_tests.py:
```diff
@@ -0,0 +1,36 @@
+import pkg_resources
+
+
+# Returns a list of PyInstaller tests.
```
Now `# Returns a list of tests which are part of the PyInstaller package.`.

bjones1 commented on PyInstaller/depend/analysis.py:
```diff
@@ -133,6 +133,68 @@ def _reset(self, user_hook_dirs):
+                    # Make sure this file exists.
+                    assert os.path.exists(abs_path), (
+                        'In %s, key %s, the file %s does not exist.' %
+                        (uhd_path, module_name, python_file_name)
```
Good point, done.

bjones1 commented on PyInstaller/depend/analysis.py:
```diff
@@ -133,6 +133,68 @@ def _reset(self, user_hook_dirs):
+            #       ...
+            #   }
+            #
+            # Check then merge this data structure, updating the file names to be absolute.
```
Hmmm. This is an interesting point. If we issue a warning, that implies (at least to me) that the current data structure in `rthooks.dat` should be `Dict[str, str]`, not `Dict[str, List[str]]`, since there shouldn't be two hooks for a given module. Put another way, why does the current data structure provide a list of files names associated with a given module? Did we at some point expect to allow multiple rthooks for each module?

bjones1 replied:
One answer to my question above: the current `rthooks.dat` specifies multiple hooks for a given module (matplotlib). The new case is when multiple `rthooks.dat` files specify hooks for the same module.

- This would be OK if the hooks don't interfere. For example, one hook for a hypothetical satellite_tracking package would apply to the load_satellites module, including its geostationary satellite data, while another package would also include a hook for load_satellites to include low earth orbit satellites.
- This would fail if the hooks conflicted. If one satellite_tracking hook specified the value for g more precisely for earth, while another specified martian gravity, one would lose.

However, I don't see how PyInstaller could decide which case applies, It could potentially issue a warning if there are multiple run-time hooks added by differing `rthooks.dat` files, but this would take additional effort (recording the `rthooks.dat` file from which each hook was added, and changing the data structure to accommodate this additional information). Therefore, I suggest omitting this check.

bjones1 commented:
In answer to @htgoebel's questions above:

> #2692 (hooks into separate repo) was closed with "won't fix', since nobody did it. Thus I suggest to keep the hooks within PyInstaller for now. These can be split out later.

Agreed.

> Example package/repository hooking into the entry points: I'd like to create some extra example package showing how to implement and test hooks. This should cover hooks, run-time hooks, pre-safe-import-module and pre-find-module-path (the later can/should be dummies, just to show how it would be done). It should also include simple testing (.travis.ci.yml).

Agreed. I'll submit PRs to your https://github.com/pyinstaller/hooksample.

> Of course this needs to be documented in the manual.

Agree; I'll do this in https://github.com/pyinstaller/hooksample as well.

You've provided a good start with https://github.com/pyinstaller/hooksample. I will extend / enhance this for all hook varieties (pre-safe, 

> I assume PyInstaller/utils/conftest.py is a 1:1-copy of tests/functional/conftest.py

Correct (mostly -- the rename is in 41dcf134d1c4d77b0aaadafa4d066ef982be6ce8), with a bit of revision to make it pull test support files from the directory where each test is located, instead of assuming a location relative to `conftest.py` -- see 155d12ac2619b67bf640a87f232c495e4e43c5b0.

> Why has test_django be moved?

This move is in 57bc96a3fcc60aedd8c9fe384f7447104111b58d. It's due to the above changes (location of support files); rather than move all the Django support files, it was easier to move `test_django.py` so that the support files would remain in their current location.

bjones1 commented:
I've updated this PR to address the concerns raised above. Important changes:

1. Because PyInstaller hooks can now include run-time hooks, I removed the special case to load run-time hooks from the code and simply moved the `rthooks` directory and `rthooks.dat` file to `PyInstaller/hooks`.
2. All log output for loading hooks now prints the path to the hook file, making it easier to understand where hooks are found given that they may come from setuptools entry points.
3. CI tests now install https://github.com/pyinstaller/hooksample, then dynamically discover and run the test therein. This shows that *any* installed package which registers tests for PyIntstaller will automatically be run.

I'm ready for a second review from @htgoebel. The next task is to provide documentation, which will be placed in https://github.com/pyinstaller/hooksample.

Other than that, this PR is done.

bjones1 commented:
OK, this is now ready. Test failures are lint; those make sense to fix up after review.

I've written a first pass of the documentation, available from https://github.com/bjones1/hooksample. To run tests and build the docs:

```
python -m pip install -U pip
pip install -e .[test,docs]
python -m PyInstaller.utils.run_tests --include_only pyi_hooksample.
sphinx-build -d _build/doctrees . _build
```

bjones1 commented:
Docs are now available at https://pyinstaller-sample-hook.readthedocs.io.

bjones1 commented:
I'll get to this when I can -- I'm still deep in the process of re-tools my classes for online only instruction.

Legorooj commented on PyInstaller/depend/analysis.py:
```diff
@@ -124,7 +124,10 @@ def _reset(self, user_hook_dirs):
-        self._user_hook_dirs = user_hook_dirs
+        # Prepend PyInstaller hook dir to user hook dirs.
+        self._user_hook_dirs = (
+            [os.path.join(PACKAGEPATH, 'hooks')] + list(user_hook_dirs)
```
I think PyInstaller hooks should be prioritised, but as #4673 discusses, have most hooks in that theoretical package - which are technically user hooks. That means PyInstaller can have very few builtin hooks, (eg PyQt5; heavily integrated into PyInstaller) which are prioritised. 

Any other ideas?

Legorooj commented on PyInstaller/depend/analysis.py:
```diff
@@ -124,7 +124,10 @@ def _reset(self, user_hook_dirs):
         self._additional_files_cache = AdditionalFilesCache()
-        self._user_hook_dirs = user_hook_dirs
+        # Prepend PyInstaller hook dir to user hook dirs.
+        self._user_hook_dirs = (
```
I agree we should rename - and I think the functional changes only should be a bit more flexible - as in we accept them assuming there's a good reason for the change.

Legorooj commented:
As to @htgoebel's Q:

> Since you are short in time: ...

I'll be submitting PRs to @bjones1 branch, which will, in turn, get merged into here. (Note the PRs will be for smaller changes/ones I truly and utterly understand)

Legorooj commented:
Oops - misclick.

htgoebel commented:
@bjones1 Okay, I'll finish this then. 

Only point Regarding prioritizing hooks:
As @Legorooj pointed out: If we ever implement #2692, all hooks will be user hooks. Thus if some package does provide its own hooks, indented to overwrite PyInstaller's, both will conflict. We need some means to prioritize. Thoughts?

bjones1 commented:
@htgoebel, thanks. I'm struggling to keep up with classes, sigh. I'd suggest making hooks added via command-line parameters the highest priority, followed next by hooks from packages (this PR), followed at lowest priority by hooks included with PyInstaller. My thoughts: first, gives the person running PyInstaller the ability to override any hook; then, assume package hooks are more updated than PyInstaller hooks.

htgoebel commented:
> I'd suggest making hooks added via command-line parameters the highest priority, followed next by hooks from packages (this PR), followed at lowest priority by hooks included with PyInstaller.

Agreed! Thanks for pointing to the command-line parameters, I would have missed them.

Rethinking, I'm going to finish this pull-request first, so both the hooksample and and #4673 can be finished. I just opend #4806 for this.

htgoebel commented:
So, finally this is done. Uff! Many, many thanks to @bjones1 for working on this.

bjones1 commented:
@htgoebel, thanks for finishing this off! I'm very excited to see it in PyInstaller. I think that will be a big step forward in making this more maintainable.
